### PR TITLE
Replace existing use of libxml2 library with expat library.

### PR DESCRIPTION
**Pull request by shivania2**
TDNF uses libxml2 library to support metalink. tdnf (via libsolv) already requires expat for xml parsing, so switching to expat library would get rid of libxml2 dependency and not adding additional dependencies.

### CODE REVIEW

File: plugins/metalink/utils.c [hunk 1]
```diff
@@ -487,80 +497,86 @@ TDNFMetalinkFree(
...
-    xmlPropValue = xmlGetProp(node, ATTR_NAME);
-    if (!xmlPropValue)
+    for (int i=0; elementInfo->attributes[i] ; i+=2)
+    {
+        if(!strcmp(elementInfo->attributes[i], ATTR_NAME))
+        {
+           name = elementInfo->attributes[i+1];
+           break;
+        }
+    }
```
Comment (oliverkurth) [inline plugins/metalink/utils.c]: Can we make this a function? This can also be used further below.

File: plugins/metalink/utils.c [hunk 2]
```diff
@@ -487,80 +497,86 @@ TDNFMetalinkFree(
...
+        {
+           type = elementInfo->attributes[i+1];
+           if (!type)
+           {
+               dwError = ERROR_TDNF_METALINK_PARSER_MISSING_HASH_ATTR;
+               pr_err("XML Parser Error:HASH element doesn't have attribute \"type\"");
+               BAIL_ON_TDNF_ERROR(dwError);
+           }
+           break;
+        }
```
Comment (oliverkurth) [inline plugins/metalink/utils.c]: Use the search function here too.

**Review by oliverkurth — COMMENTED** [review]
Overall it looks good, thank you for the work! I'll have another look later.

**Comment by sshedi** [comment]
Good job Shivani. This was in my backlog from a long time, glad that you did it. This helps. One less package in our docker image.

File: plugins/metalink/utils.c (hunk 1 above)
Comment (shivania2) [reply plugins/metalink/utils.c]: Done

File: plugins/metalink/utils.c (hunk 2 above)
Comment (shivania2) [reply plugins/metalink/utils.c]: Done

File: plugins/metalink/utils.c [hunk 3]
```diff
@@ -485,82 +495,92 @@ TDNFMetalinkFree(
     TDNF_SAFE_FREE_MEMORY(ml_ctx);
 }
 
+char *TDNFSearchTag(const char **attr, const char *type)
+{
+
```
Comment (sshedi) [inline plugins/metalink/utils.c]: unnecessary new line

File: plugins/metalink/utils.c [hunk 4]
```diff
@@ -485,82 +495,92 @@ TDNFMetalinkFree(
...
+    char *value = NULL;
+
+    for (int i=0; attr[i] ; i+=2)
```
Comment (sshedi) [inline plugins/metalink/utils.c]: please keep the code uniform, space after attr[i] is not needed here

i = 0, i += 2

File: plugins/metalink/utils.c [hunk 5]
```diff
@@ -485,82 +495,92 @@ TDNFMetalinkFree(
...
+    {
+        if(!strcmp(attr[i], type))
+        {
+           value = (char *)attr[i+1];
```
Comment (sshedi) [inline plugins/metalink/utils.c]: attr[i+1] => attr[i + 1]

File: plugins/metalink/utils.c [hunk 6]
```diff
@@ -485,82 +495,92 @@ TDNFMetalinkFree(
     TDNF_SAFE_FREE_MEMORY(ml_ctx);
 }
 
+char *TDNFSearchTag(const char **attr, const char *type)
+{
+
+    char *value = NULL;
```
Comment (sshedi) [inline plugins/metalink/utils.c]: this variable use can be avoided in this function.

```
        if(!strcmp(attr[i], type))
        {
            return (char *) attr[i + 1];
         }
        ...
   return NULL;
}
```

File: plugins/metalink/utils.c [hunk 7]
```diff
@@ -485,82 +495,92 @@ TDNFMetalinkFree(
...
+    void *userData
     )
 {
     uint32_t dwError = 0;
-    xmlChar* xmlPropValue = NULL;
     const char *name = NULL;
+    struct MetalinkElementInfo* elementInfo = (struct MetalinkElementInfo*)userData;
 
-    if(!ml_ctx || !node || IsNullOrEmptyString(filename))
+    if(!elementInfo->ml_ctx || IsNullOrEmptyString(elementInfo->filename))
```
Comment (sshedi) [inline plugins/metalink/utils.c]:
```
if (!elementInfo || !elementInfo->ml_ctx || IsNullOrEmptyString(elementInfo->filename))
```

File: plugins/metalink/utils.c [hunk 8]
```diff
@@ -610,18 +622,19 @@ TDNFParseHashTag(
...
-    xmlChar* xmlContValue = NULL;
-    const char *value = NULL;
+    char *value = NULL;
+    char *prefval = NULL;
     int prefValue = 0;
     TDNF_ML_URL_INFO *ml_url_info = NULL;
+    struct MetalinkElementInfo* elementInfo = (struct MetalinkElementInfo*)userData;
 
-    if(!ml_ctx || !node)
+    if(!elementInfo->ml_ctx)
```
Comment (sshedi) [inline plugins/metalink/utils.c]:
```
if (!elementInfo || !elementInfo->ml_ctx)
```

File: plugins/metalink/utils.c [hunk 9]
```diff
@@ -720,128 +720,146 @@ TDNFParseUrlTag(
...
-uint32_t
-TDNFXmlParseData(
-    TDNF_ML_CTX *ml_ctx,
-    xmlNode *node,
-    const char *filename
-    )
-{
-    uint32_t dwError = 0;
-    xmlChar* xmlContValue = NULL;
+void TDNFXmlParseStartElement(void *userData, const char *name, const char **attrs){
```
Comment (sshedi) [inline plugins/metalink/utils.c]: follow the function definition style, like done in other places

File: plugins/metalink/utils.c [hunk 10]
```diff
@@ -720,128 +720,146 @@ TDNFParseUrlTag(
...
+    struct MetalinkElementInfo* elementInfo = (struct MetalinkElementInfo*)userData;
+
+    if(elementInfo->dwError != 0)
```
Comment (sshedi) [inline plugins/metalink/utils.c:733]: use curly braces while using a macro inside if condition

File: plugins/metalink/utils.c [hunk 11]
```diff
@@ -720,128 +720,146 @@ TDNFParseUrlTag(
...
+error:
+    goto cleanup;
+}
+
+void TDNFXmlParseData(void *userData, const char *val, int len){
+    struct MetalinkElementInfo* elementInfo = (struct MetalinkElementInfo*)userData;
     char *size = NULL;
 
-    if(!ml_ctx || !node || IsNullOrEmptyString(filename))
+    if(!elementInfo->ml_ctx || IsNullOrEmptyString(elementInfo->filename) || (elementInfo->dwError != 0))
```
Comment (sshedi) [inline plugins/metalink/utils.c]: null check for elementInfo

File: plugins/metalink/utils.c [hunk 12]
```diff
@@ -720,128 +720,146 @@ TDNFParseUrlTag(
...
-        BAIL_ON_TDNF_ERROR(dwError);
-    }
+    XML_SetElementHandler(parser, TDNFXmlParseStartElement, TDNFXmlParseEndElement);
+    XML_SetCharacterDataHandler(parser, TDNFXmlParseData);
+    XML_SetUserData(parser, &elementInfo);
 
-    //Get the root element from parsed xml tree.
-    root_element = xmlDocGetRootElement(doc);
+    // Determine the file size
+    fseek(file, 0, SEEK_END);
```
Comment (sshedi) [inline plugins/metalink/utils.c]: validate fseek ret , you can also use `stat(...)` to get file size

**Review by sshedi — CHANGES_REQUESTED** [review]

File: plugins/metalink/utils.c (hunk 3 above)
Comment (shivania2) [reply plugins/metalink/utils.c]: Done

File: plugins/metalink/utils.c (hunk 4 above)
Comment (shivania2) [reply plugins/metalink/utils.c]: done

File: plugins/metalink/utils.c (hunk 5 above)
Comment (shivania2) [reply plugins/metalink/utils.c]: done

File: plugins/metalink/utils.c (hunk 6 above)
Comment (shivania2) [reply plugins/metalink/utils.c]: Done

File: plugins/metalink/utils.c (hunk 7 above)
Comment (shivania2) [reply plugins/metalink/utils.c]: Done

File: plugins/metalink/utils.c (hunk 8 above)
Comment (shivania2) [reply plugins/metalink/utils.c]: Done

File: plugins/metalink/utils.c (hunk 9 above)
Comment (shivania2) [reply plugins/metalink/utils.c]: Done

File: plugins/metalink/utils.c (hunk 12 above)
Comment (shivania2) [reply plugins/metalink/utils.c]: Used fstat() to determine file size

File: plugins/metalink/utils.c (hunk 11 above)
Comment (shivania2) [reply plugins/metalink/utils.c]: Done

File: plugins/metalink/utils.c (hunk 10 above)
Comment (shivania2) [reply plugins/metalink/utils.c:733]: Done

**Review by oliverkurth — APPROVED** [review]
LGTM. I'll merge if others agree.

**Review by sshedi — APPROVED** [review]

File: plugins/metalink/utils.c [hunk 13]
```diff
@@ -631,83 +644,70 @@ TDNFParseUrlTag(
...
+    for (int i = 0; elementInfo->attributes[i]; i += 2)
     {
-        value = (const char*)xmlPropValue;
-        if(sscanf(value, "%d", &prefValue) != 1)
+        if (!strcmp(elementInfo->attributes[i], ATTR_PROTOCOL))
         {
-            dwError = ERROR_TDNF_INVALID_PARAMETER;
-            pr_err("XML Parser Warning: Preference is invalid value: %s\n", value);
-            BAIL_ON_TDNF_ERROR(dwError);
+           value = (char *)elementInfo->attributes[i + 1];
```
Comment (ntsbtz) [inline plugins/metalink/utils.c:652]: isnt this going to crash if attributes key is present but value is missing, Its better to check atributes[i+1] is not NULL.If taking care of this please do it for wherever using this atributes[i+1]  directly.

File: plugins/metalink/utils.c [hunk 14]
```diff
@@ -720,128 +720,171 @@ TDNFParseUrlTag(
...
+            BAIL_ON_TDNF_ERROR(elementInfo->dwError);
         }
-        node = node->next;
+    }
+    else if(!strcmp(elementInfo->startElement, TAG_NAME_HASH))
+    {
+        elementInfo->dwError = TDNFParseHashTag(userData, val, len);
+        BAIL_ON_TDNF_ERROR(elementInfo->dwError);
+    }
+    else if(!strcmp(elementInfo->startElement, TAG_NAME_URL) && len > 4)
```
Comment (ntsbtz) [inline plugins/metalink/utils.c]: Why its 4 not any other number. Better to replace with a meaningful MACRO.

File: plugins/metalink/utils.c [hunk 15]
```diff
@@ -720,128 +720,171 @@ TDNFParseUrlTag(
...
     TDNF_ML_CTX *ml_ctx,
-    int fd,
+    FILE *file,
     const char *filename
     )
 {
     uint32_t dwError = 0;
-    xmlDoc *doc = NULL;
-    xmlNode *root_element = NULL;
+    struct stat st;
```
Comment (ntsbtz) [inline plugins/metalink/utils.c]: Its better to initialise variables .

File: plugins/metalink/utils.c [hunk 16]
```diff
@@ -720,128 +720,171 @@ TDNFParseUrlTag(
...
-    doc = xmlReadFd(fd, NULL, NULL, 0);
+    elementInfo.ml_ctx = ml_ctx;
+    elementInfo.filename = filename;
 
-    if (doc == NULL)
-    {
-        pr_err("%s: Error while reading xml from fd:%d \n", __func__, fd);
-        dwError = ERROR_TDNF_METALINK_PARSER_INVALID_DOC_OBJECT;
-        BAIL_ON_TDNF_ERROR(dwError);
+    XML_SetElementHandler(parser, TDNFXmlParseStartElement, TDNFXmlParseEndElement);
```
Comment (ntsbtz) [inline plugins/metalink/utils.c:848]: These function has return type of XMLCALL. Please check once that is void or something else so can check if function failed.

**Review by ntsbtz — CHANGES_REQUESTED** [review]
Have some minor comments and query. Please look into that and rest is LGTM!!!

File: plugins/metalink/utils.c (hunk 13 above)
Comment (shivania2) [reply plugins/metalink/utils.c:652]: Done

File: plugins/metalink/utils.c (hunk 14 above)
Comment (shivania2) [reply plugins/metalink/utils.c]: Length 4 is used as minimum url length. Added macro for this number

File: plugins/metalink/utils.c (hunk 15 above)
Comment (shivania2) [reply plugins/metalink/utils.c]: Done

File: plugins/metalink/utils.c (hunk 16 above)
Comment (shivania2) [reply plugins/metalink/utils.c:848]: XML_SetElementHandler has return type void

```
void XMLCALL
XML_SetElementHandler(XML_Parser p,
                      XML_StartElementHandler start,
                      XML_EndElementHandler end);
```

File: plugins/metalink/utils.c [hunk 17]
```diff
@@ -720,128 +721,172 @@ TDNFParseUrlTag(
...
-    if(!ml_ctx || (fd <= 0) || IsNullOrEmptyString(filename))
+    struct stat st;
+    size_t file_size = 0;
+    char *buffer = NULL;
+    struct MetalinkElementInfo elementInfo;
+    elementInfo.startElement = NULL;
+    elementInfo.endElement = NULL;
+    elementInfo.dwError = 0;
+
+    memset(&st, 0, sizeof(struct stat));
```
Comment (ntsbtz) [inline plugins/metalink/utils.c]: Replace this with  st.st_size = -1 as we dont need whole structure info.
Except this LGTM!!

**Review by ntsbtz — APPROVED** [review]

File: plugins/metalink/utils.c (hunk 17 above)
Comment (shivania2) [reply plugins/metalink/utils.c]: Done

File: plugins/metalink/utils.c [hunk 18]
```diff
@@ -720,128 +721,172 @@ TDNFParseUrlTag(
...
+    FILE *file,
     const char *filename
     )
 {
     uint32_t dwError = 0;
-    xmlDoc *doc = NULL;
-    xmlNode *root_element = NULL;
-
-    if(!ml_ctx || (fd <= 0) || IsNullOrEmptyString(filename))
+    struct stat st;
```
Comment (oliverkurth) [inline plugins/metalink/utils.c]: Use `struct stat st = {0}` - this would initialize all elements to 0.

File: plugins/metalink/utils.c [hunk 19]
```diff
@@ -720,128 +721,172 @@ TDNFParseUrlTag(
...
+    size_t file_size = 0;
+    char *buffer = NULL;
+    struct MetalinkElementInfo elementInfo;
+    elementInfo.startElement = NULL;
+    elementInfo.endElement = NULL;
+    elementInfo.dwError = 0;
+    st.st_size = 0;
```
Comment (oliverkurth) [inline plugins/metalink/utils.c]: If you use `{0}` as suggested above this wont be needed.

File: plugins/metalink/utils.c [hunk 20]
```diff
@@ -720,128 +721,172 @@ TDNFParseUrlTag(
...
 {
     uint32_t dwError = 0;
-    xmlDoc *doc = NULL;
-    xmlNode *root_element = NULL;
-
-    if(!ml_ctx || (fd <= 0) || IsNullOrEmptyString(filename))
+    struct stat st;
+    size_t file_size = 0;
+    char *buffer = NULL;
+    struct MetalinkElementInfo elementInfo;
```
Comment (oliverkurth) [inline plugins/metalink/utils.c]: Also here: `struct MetalinkElementInfo elementInfo = {0}`

File: plugins/metalink/utils.c (hunk 18 above)
Comment (shivania2) [reply plugins/metalink/utils.c]: Done

File: plugins/metalink/utils.c (hunk 19 above)
Comment (shivania2) [reply plugins/metalink/utils.c]: Done

File: plugins/metalink/utils.c (hunk 20 above)
Comment (shivania2) [reply plugins/metalink/utils.c]: Done

**Pull request merged**